crates/rpc/rpc-engine-api/src/engine_api.rs

```diff
@@ -62,6 +90,35 @@ impl<Client: HeaderProvider + BlockProvider + StateProvider> EngineApi<Client> {
         }
     }
 
+    /// Validates the presence of the `withdrawals` field according to the payload timestamp.
+    /// After Shanghai, withdrawals field must be [Some].
+    /// Before Shanghai, withdrawals field must be [None];
+    fn validate_withdrawals_presence(
+        &self,
+        version: EngineApiMessageVersion,
+        timestamp: u64,
+        has_withdrawals: bool,
+    ) -> EngineApiResult<()> {
+        let is_shanghai = self.chain_spec.fork(Hardfork::Shanghai).active_at_timestamp(timestamp);
+
+        match version {
+            EngineApiMessageVersion::V1 => {
+                if is_shanghai || has_withdrawals {
+                    return Err(EngineApiError::InvalidParams)
+                }
+            }
+            EngineApiMessageVersion::V2 => {
+                let shanghai_with_no_withdrawals = is_shanghai && !has_withdrawals;
+                let not_shanghai_with_withdrawals = !is_shanghai && has_withdrawals;
+                if shanghai_with_no_withdrawals || not_shanghai_with_withdrawals {
```

Comment:
what about the "no shanghai no withdrawals" case? do we allow V2 messages then?

Reply:
yes, we do. this is similar to consensus check where withdrawals **must** be present after shanghai and **must** be absent before